roslaunch_editor/www/main.js

```diff
@@ -0,0 +1,287 @@
+		} else if (type == 'int') {
+			html += `<input data-item="${esc(item)}" data-path="${path}" type=number value="${esc(value)}">`;
+		} else if (type == 'float') {
+			let step = '0.1';
+			try { step = '0.' + '0'.repeat(value.match(/\.(.*)$/)[1].length - 1) + '1'; } catch {}; // TODO: scientific
+			html += `<input data-item="${esc(item)}" data-path="${path}" type=number step="${step}" value="${esc(value)}">`;
+		} else if (options) {
+			if (!options.includes(value)) {
+				options.unshift(value);
+			}
+			var optionsHTML = options.map(function(option) {
+				return `<option${option == value ? ' selected' : ''}>${esc(option)}</option>`;
+			}).join('');
+			html += `<select data-item="${esc(item)}" data-path="${path}">${optionsHTML}</select>`;
+		} else {
+			html += `<input data-item="${esc(item)}" data-path="${path}" value="${esc(value)}">`;
+		}
+
+		html += `<span class=comment title="${esc(comment)}">${esc(comment)}</span></label>`
+	}
+
+	return html;
+}
+
+function updateDocs() {
+	editorElem.querySelectorAll('input, select').forEach(function(elem) {
+		var type = elem.getAttribute('type');
+		if (type == 'checkbox') {
+			var value = String(elem.checked);
+		} else {
+			var value = elem.value;
+		}
+		var path = elem.getAttribute('data-path');
+		var item = elem.getAttribute('data-item');
+		var element = items[item].doc.querySelector(path);
+		if (element.getAttribute('default') === null && value == '') {
+			// don't add empty default if it's not set
+			return;
+		}
+		element.setAttribute('default', value);
+	});
+}
+
+editorElem.addEventListener('change', updateDocs);
+
+function addLaunchFile(name, content) {
+	html = `<fieldset><legend><code>${name}</code></legend>`;
+	html += generateForm(name, content);
+	html += '</fieldset>'
+	editorElem.innerHTML += html;
+}
+
+function parseItem(str) {
+	var parsed = str.match(/(.*?)\/(.*)/);
+	var package = parsed[1];
+	var name = parsed[2];
+	return items[str] = { package, name };
+}
+
+var clover;
+
+function determineMode() {
+	return new Promise(function(resolve, reject) {
+		function errcb(err) {
+			alert(err);
+			reject();
+		}
+		// check roslaunch_editor's read service
+		ros.getServiceType('/roslaunch_editor/read', function(t) {
+			if (t == 'roslaunch_editor/ReadLaunchFiles') {
+				clover = false;
+				resolve();
+				return;
+			}
+			// check clover's exec service
+			ros.getServiceType('/exec', function(t) {
+				if (t == 'clover/Execute') {
+					clover = true;
+					resolve();
+					return;
+				}
+				alert('Neither /roslaunch_editor/read nor /exec service not found');
+				reject();
+			}, errcb);
+		}, errcb);
+	});
+}
+
+function errCallback(err) {
+	alert('Error calling service: ' + err);
+}
+
+function loadItems() {
+	editorElem.innerHTML = '';
+
+	if (typeof p.items == 'string') {
+		p.items = p.items.split(',');
+	}
+
+	if (clover) {
+		const boundary = '===BOUNDARY===';
+		var cmd = 'bash -ic "' + p.items.map(function(item) {
```

Comment:
Calling `bash` from JS seems... odd, and the other branch looks much cleaner. Just sayin'.